### PR TITLE
Add FIO naming support (testnet)

### PR DESCRIPTION
**Pull request by catenocrypt**
Extending naming providers with FIO naming, FIO testnet for now.  Issue #760 .
Format:  'name@fiotestnet'; API endpoint: http://testnet.fioprotocol.io/v1/chain

For testing:
make functional

```
curl http://localhost:8420/v2/ns/lookup?name=adam@fiotestnet\&coins=60
[{"result":"0xce5cB6c92Da37bbBa91Bd40D4C9D4D724A3a8F51","coin":60}]
```

### CODE REVIEW

File: api/naming_service.go [hunk 1]
```diff
@@ -122,3 +123,21 @@ func handleLookup(name string, coins []uint64) (result []blockatlas.Resolved, er
 	result, err = api.Lookup(coins, name)
 	return
 }
+
+// Obtain tld from then name, e.g. ".ens" from "nick.ens"
+func getTLD(name string) (tld string, ok bool) {
```
Comment (vikmeup) [inline api/naming_service.go]: Test?

File: platform/fio/api.go [hunk 2]
```diff
@@ -22,3 +22,20 @@ func (p *Platform) Coin() coin.Coin {
...
+		if err != nil {
+			return result, err
+		}
+		result = append(result, blockatlas.Resolved{Coin: coinId, Result: address})
+	}
+
+	return result, nil
+}
+
+
```
Comment (vikmeup) [inline platform/fio/api.go]: Did you run formatting and linting?

File: platform/fio/client.go [hunk 3]
```diff
@@ -2,8 +2,32 @@ package fio
...
 import (
 	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+	"github.com/trustwallet/blockatlas/pkg/errors"
 )
 
 type Client struct {
 	blockatlas.Request
 }
+
+type GetPubAddressRequest struct {
```
Comment (EnoRage) [inline platform/fio/client.go]: Maybe it is ok to add structures to the model.go file like any other platform?
[example](https://github.com/trustwallet/blockatlas/blob/master/platform/cosmos/model.go)

File: platform/fio/api.go [hunk 4]
```diff
@@ -22,3 +22,20 @@ func (p *Platform) Coin() coin.Coin {
 func (p *Platform) GetTxsByAddress(address string) (page blockatlas.TxPage, err error) {
 	return page, err
 }
+
+func (p *Platform) Lookup(coins []uint64, name string) ([]blockatlas.Resolved, error) {
+	var result []blockatlas.Resolved
+	for _, coinId := range coins {
+		coinObj := coin.Coins[uint(coinId)]
+		address, err := p.client.lookupPubAddress(name, coinObj.Symbol)
+		if err != nil {
```
Comment (EnoRage) [inline platform/fio/api.go]: Maybe we can ignore coins with errors and return the correct ones
If I will have an array of coins where only 1 is incorrect - it is better to return all the info except the invalid coin?

File: api/naming_service.go [hunk 5]
```diff
@@ -122,3 +123,21 @@ func handleLookup(name string, coins []uint64) (result []blockatlas.Resolved, er
...
+	if strings.Index(name, ".") >= 0 {
+		// xxx.tld format
+		ss := strings.Split(name, ".")
+		tld := "." + ss[len(ss)-1]
+		return tld, true
+	}
+	if strings.Index(name, "@") >= 0 {
+		// xxx@tld format
```
Comment (hewigovens) [inline api/naming_service.go]: Is it possbile to have something like: xxx.ens@fio?

File: platform/fio/client.go [hunk 6]
```diff
@@ -2,8 +2,32 @@ package fio
...
+
+type GetPubAddressResponse struct {
+	PublicAddress 	string  `json:"public_address"`
+	BlockNum  		int 	`json:"block_num"`
+	Message         string  `json:"message"`
+}
+
+func (c *Client) lookupPubAddress(name string, coinSymbol string) (address string, error error) {
+	var res GetPubAddressResponse
+	err := c.Post(&res, "get_pub_address", GetPubAddressRequest{FioAddress: name, TokenCode: coinSymbol})
```
Comment (hewigovens) [inline platform/fio/client.go]: does the api endpoint support batching?

File: api/naming_service.go (hunk 1 above)
Comment (catenocrypt) [reply api/naming_service.go]: unit test added

File: platform/fio/api.go (hunk 2 above)
Comment (catenocrypt) [reply platform/fio/api.go]: formatting done

File: platform/fio/client.go (hunk 3 above)
Comment (catenocrypt) [reply platform/fio/client.go]: Done, thanks

File: api/naming_service.go (hunk 5 above)
Comment (catenocrypt) [reply api/naming_service.go]: TLD separation logic reworked, this special case is handled now, unit test added.

File: platform/fio/client.go (hunk 6 above)
Comment (catenocrypt) [reply platform/fio/client.go]: I don't think so.  Not a typical use case.

File: platform/fio/api.go (hunk 4 above)
Comment (catenocrypt) [reply platform/fio/api.go]: Good suggestion.  This is a matter of taste/usage pattern.  After some consideration I think in partial error case the caller is more likely to handle error if error is returned than if partial error is returned.  That has the risk of not noting error, and attempting to use missing address.  Better to return error if not all could be resolved.  Typical use case is with one coin anyways.

File: platform/fio/client.go [hunk 7]
```diff
@@ -2,8 +2,21 @@ package fio
...
 
 type Client struct {
 	blockatlas.Request
 }
+
+func (c *Client) lookupPubAddress(name string, coinSymbol string) (address string, error error) {
+	var res GetPubAddressResponse
+	err := c.Post(&res, "get_pub_address", GetPubAddressRequest{FioAddress: name, TokenCode: coinSymbol})
+	if err != nil {
+		return "", err
```
Comment (Pantani) [inline platform/fio/client.go]:
```suggestion
		return "", errors.E(err, "Error lokking up FIO name", errors.Params{"name": name, "coinSymbol": coinSymbol})
```

File: platform/fio/client.go [hunk 8]
```diff
@@ -2,8 +2,21 @@ package fio
...
+	}
+	if res.Message != "" {
+		return "", errors.E("Error lokking up FIO name: "+res.Message, errors.Params{"name": name, "coinSymbol": coinSymbol})
```
Comment (Pantani) [inline platform/fio/client.go]:
```suggestion
		return "", errors.E("Error lokking up FIO name", errors.Params{"name": name, "coinSymbol": coinSymbol, "Message": res.Message})
```

File: api/naming_service.go [hunk 9]
```diff
@@ -122,3 +124,17 @@ func handleLookup(name string, coins []uint64) (result []blockatlas.Resolved, er
 	result, err = api.Lookup(coins, name)
 	return
 }
+
+// Obtain tld from then name, e.g. ".ens" from "nick.ens"
+func getTLD(name string) (tld string, ok bool) {
```
Comment (dpereskokov) [inline api/naming_service.go]: may be better to return `err` instead of `ok`?

**Pull request merged**